zerver/lib/markdown/__init__.py

```diff
@@ -1756,9 +1756,8 @@ def run(self, lines: List[str]) -> List[str]:
                 fence_str = m.group("fence")
                 lang: Optional[str] = m.group("lang")
                 is_code = lang not in ("quote", "quoted")
-                has_open_fences = not len(open_fences) == 0
                 matches_last_fence = (
-                    fence_str == open_fences[-1].fence_str if has_open_fences else False
+                    fence_str == open_fences[-1].fence_str if open_fences else False
```

Comment:
I suppose mypy helps somewhat here, but as a general rule, I prefer `len(open_fences) > 0` here over using the falsiness of lists, because there's other ways a variable can be falsey (say, being `None`), and I think it's better to be explicit about what we mean.

(I'd probably be happiest just changing the `not len(open_fences) == 0` to just `len(open_fences) > 0` here, and keeping the variable name).

I guess the downside of that proposal is that the current code is quadratic, and the new code might not be.

Reply:
We should definitely avoid relying on falsiness of variables that could be `None`. But that’s not this case; this is a textbook case from [PEP 8](https://peps.python.org/pep-0008/#programming-recommendations).

> For sequences, (strings, lists, tuples), use the fact that empty sequences are false:
>
> ```python
> # Correct:
> if not seq:
> if seq:
> ```
>
> ```python
> # Wrong:
> if len(seq):
> if not len(seq):
> ```

I hadn’t looked closely at the algorithm because I didn’t want to throw up on the plane. If nothing else, it violates CommonMark: “The content of the code block consists of all subsequent lines, until a closing [code fence](https://spec.commonmark.org/0.30/#code-fence) of the same type as the code block began with (backticks or tildes), and with **at least as many** backticks or tildes as the opening code fence.”